### ISSUE TITLE
[feature request] SemVer auto versioning possible?

### ISSUE THREAD

**Issue by Thaina** [NuGet/Home#8371]
I think more and more people understanding versioning number in SemVer format to publish with nuget. And SemVer is mostly straightforward. It seem like we could possibly have analyzer and diff automatically with C# reflection by comparing public signature & deprecate attribute & file hash

- If all public signature is the same but file hash changed, bump the patch
- If there are any new public signature | deprecate attribute then bump the minor
- If there are any missing public signature (include change of signature) then bump the major

And so I wish that we would have a command to let nuget analyze the build number from SemVer rule compare with previous published version. Or if possible, I wish that nuget would bump the version in project config file comparing to previous git commit whenever the project built

**Comment by xavierdecoster** [comment]
SemVer is technology agnostic and not restricted to versioning what is found in code. It's about stating intentions as an owner of the API, and communicating this to the API consumer in a semantic way.

Any automated diff tool that scans the code base for changes in public signatures may merely suggest a new version number based on the scan results. In the end, IMHO it should still be a conscious decision by the API (or package) author who defines and controls that public API (which may be more than just the code).

See [my blog post](https://www.xavierdecoster.com/post/2014/03/24/semantic-versioning-state-your-intentions-please) from 2014 where I explain this thought.

**Comment by Thaina** [comment]
@xavierdecoster I think it should also considered that we could `intend to make it auto` with the project config

I just think that we should have an option in project config (`"autoSemver" : true` maybe? let it false by default) and nuget will hook its semver analyzer in post build event

We might also put other process to let we have clearer intention. But it should be the default process that it would always detect that we should not forget to change version when it really should be. Such when we have change the signature of function, not just add new overload. It must bump the major without the need for us to always remember

And for other things that's not a code then sure we need to specify it manually. I think we could just let the process can be manually override but still the process will stay

Such as, if we need put the version manually, the analyzer will respect our manual version that has more value
Suppose we have commit version `1.0.1`
Then we build normally, it check with source control and put version `1.0.2`
But if we have made some non code minor change so we manually specify `1.1.1`, if we just make any other minor change in code the analyzer will let it as is
But then if we make a breaking change in code, it will try to override to `2.0.0`, something like that

**Comment by Thaina** [comment]
@xavierdecoster I have fully read your blog post and I have disagree with you at one thing

You don't need to `see standard definition of public API`

Because you could shape it by your hand and let us have discussion here

I almost agree with your definition I have seen. And just let us manually override case by case. But your definition would serve as a base, a fallback, a failsafe. So we would commit and publish with more convenient with SemVer. At least we have the most lowest safeguard to publish any version and will not likely break most of other people repo who loyal to SemVer semantic dependency

**Comment by xavierdecoster** [comment]
I do understand what you're after, but I don't think it's a NuGet responsibility. It's analysis that should happen before creating the NuGet package.

Here's some additional literature that might convince you that there's friction, a contradiction if you will, which makes it impossible for automation to *decide* the next version. It can merely analyze and suggest, and *help the author to decide*. I agree that such tooling would be useful to assist package authors in respecting versioning semantics and not to forget about that breaking API change for instance.

* https://devblogs.microsoft.com/devops/versioning-nuget-packages-cd-1/ (do read the follow-up post as well)
* https://www.xavierdecoster.com/post/2013/04/29/semantic-versioning-auto-incremented-nuget-package-versions (old post, but it explains the paradox of **auto-versioning a yet unknown semantic version**)
* @HowardvanRooijen wrote an [article](https://blogs.endjin.com/2016/02/an-experiment-to-automatically-detect-api-breaking-changes-in-dot-net-assemblies-and-suggest-a-semantic-version-number/) about a cool experiment he did and [published the code](https://github.com/endjin/Endjin.Assembly.ChangeDetection) to GitHub. Maybe you want to have a play with it and perhaps it can help you accomplish what you're after.

Hope that helps!

**Comment by Thaina** [comment]
@xavierdecoster I think I have mostly agree with you. But my point is, the process of `analyze and suggest, and help the author to decide` could be happen at build time (not a publishing time). And the result is just putting the version number directly in project config for nuget (because this is where we would want to have SemVer semantic be utilized)

At the time we would publish. If there was anything we want to manual deciding, we could just manually did it before publish or even before that. And it would loyal to our manual decision unless the analyzer found something more serious

As I said I hope this feature is a failsafe. And actually I was thinking about this failsafe when I have specify SemVer semantic for CD. I have propose this feature in nuget because, as I said above, nuget is where we would want to have SemVer semantic be utilized. We use nuget for CD and specify SemVer wildcard all around the place. One wrong publish would break all the chained project. And so at least guarding against code breaking would be very precious

**Comment by Thaina** [comment]
@xavierdecoster I have try to understand the friction between SemVer and nuget site. But then I realize that it opposite

If we just try to let SemVer be standardized then the nuget would have more reason to adopt it in the system

And for all of those. The concept of `major.minor.patch` could be the same for any versioning kind. semantic string or build number is not the important point. I think even if we made the version compatible with C# version object (it cannot contain any string. must be only 4 interger to parse) the 3 first number could still be use the SemVar logic too

**Comment by xavierdecoster** [comment]
NuGet fully supports the SemVer2 spec nowadays and simply accepts whatever version you want to use for your package. There are reasons why SemVer evolved from v1 to v2 and introduced support for build metadata, so narrowing it down again and only support integers would be a mistake.

There are multiple issues with the feature you request that are out of scope for NuGet and should be handled before even invoking our tools to create the package.

In my experience, if a whole chain of packages breaks because of an error in version number, then the issue likely is not versioning, or even tooling, but the way the binaries are componentized or how tightly coupled they are.

I'd also like to point out that, even though SemVer is a good strategy for many reasons, it's not a silver bullet either. All we can do is support it for those who are willing to use it. We can recommend people to use SemVer for redistributable binaries to communicate intent, but we cannot and should not enforce it. Simply put: it's not up to us to _guess_ or _decide_ what the package author's intentions are with each new package version.

**Comment by Thaina** [comment]
@xavierdecoster Firstly. I think it's not that we would `narrow it down` to 3 integers. But I just scope this feature/tool to focus on only first 3 integer. And we don't need to care for the last part, be it integer or metadata string is up to each developer decision

But the main concern is, if there would be any change that should be seriously concern. Especially adding or removing public api. There should be a way to never forget to bump the version as SemVer intended to be

Maybe we could have opposite, having a publish guard that analyze and make the publish command fail if the publisher should bump the version but did not, and suggest there

Second. Suppose we have API like this in some package

```C#
/// package version 1.0.0
public void A(long l) => DoSomething(l);
```

And make CD with this package with version 1.0.* and call this function
```C#
long l = 10e8;
A(l);
```
Then the owner have update some minor things but also change this function into this

```C#
/// package version 1.0.1
public void A(int i) => DoSomething(i);
```

This would break my CD build and need me to go fix it. This change actually should be version 2.0.0 and if we have analyzer we could avoid this human error easily

**Comment by nkolev92** [comment]
Thanks for filing this @Thaina, but I agree with @xavierdecoster, that this is not the client's responsibility. 
The client supports semver, but does not enforce it.
We should not/cannot make assumptions about the user's intentions.

There are too many technical challenges and scope creeps with NuGet meddling with versions at build time. 

While I understand efforts by other tools(or package feeds, 3rd party or even nuget.org) to satisfy your scenario, I don't believe that has a place in the core client.

**Comment by nkolev92** [comment]
Closing per the above comment.
